planet/api/http.py

```diff
@@ -13,135 +13,176 @@
-#
 
-def _log_request(req):
-    LOGGER.info('%s %s %s %s', req.method, req.url, req.params, req.data)
+class Session():
+    '''Context manager for asynchronous communication with the Planet server.
 
+    Authentication for Planet servers is given as ('<api key>', '').
 
-class PlanetSession():
-    """Provides communication with the Planet server"""
+    :param auth: Planet server authentication.
+    :type auth: httpx.Auth or tuple.
+    '''
 
-    def __init__(self):
-        # general session for sync api calls
-        self._session = RedirectSession()
-        self._session.headers.update({'User-Agent': self._get_user_agent()})
-        self._session.verify = USE_STRICT_SSL
+    def __init__(self, auth=None):
+        self._client = httpx.AsyncClient(auth=auth)
+        self._client.headers.update({'User-Agent': self._get_user_agent()})
+        self._client.event_hooks['request'] = [self._log_request]
+        self._client.event_hooks['response'] = [
+            self._log_response,
+            self._raise_for_status
+        ]
         self.retry_wait_time = RETRY_WAIT_TIME
+        self.retry_count = RETRY_COUNT
 
-    @staticmethod
-    def _get_user_agent():
-        return 'planet-client-python/' + __version__
-
-    def __enter__(self):
+    async def __aenter__(self):
         return self
 
-    def __exit__(self, *args):
-        self.close()
+    async def __aexit__(self, *args):
+        await self.aclose()
 
-    def close(self):
-        self._session.close()
+    async def aclose(self):
+        await self._client.aclose()
 
-    def request(self, request, retry_count=5):
-        '''Submit a request with retry.
+    async def retry(self, func, *a, **kw):
+        '''Run an asynchronous request function with retry.'''
+        # TODO: retry will be provided in httpx v1 [1] with usage [2]
+        # 1. https://github.com/encode/httpcore/pull/221
+        # 2. https://github.com/encode/httpx/blob/
+        # 89fb0cbc69ea07b123dd7b36dc1ed9151c5d398f/docs/async.md#explicit-transport-instances # noqa
+        # TODO: if throttling is necessary, check out [1] once v1
+        # 1. https://github.com/encode/httpx/issues/984
+        retry_count = self.retry_count
+        wait_time = self.retry_wait_time
 
-        :param :py:Class:`planet.api.models.Request` req: request to submit
-        :param int retry_count: number of retries
-        :returns: :py:Class:`planet.api.models.Response`
-        '''
         max_retry = retry_count + 1
         for i in range(max_retry):
             try:
-                resp = self._do_request(request)
-                return resp
+                return await func(*a, **kw)
             except exceptions.TooManyRequests:
                 if i < max_retry:
                     LOGGER.debug(f'Try {i}')
-                    LOGGER.info('Too Many Requests: sleeping and retrying')
+                    LOGGER.info(f'Too Many Requests: sleeping {wait_time}s')
```

Comment:
This doesn't seem to provide enough information for a user to being able to dig in to which requests are being throttled. Consider adding more identifying information like verb, URI, and/or a temporary (shortened) SHA to provide more filtering criteria on these log entries.

Reply:
no, I'm fine with `TooManyRequests` just saying "yo dudes too many requests". If a dev wants more info they can wrap their own call with logging.